optuna/integration/pytorch_lightning.py

```diff
@@ -52,7 +68,39 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
             warnings.warn(message)
             return
 
-        self._trial.report(current_score, step=epoch)
-        if self._trial.should_prune():
+        should_stop = False
+        if trainer.is_global_zero:
+            self._trial.report(current_score, step=epoch)
+            should_stop = self._trial.should_prune()
+        should_stop = trainer.training_type_plugin.broadcast(should_stop)
+        if not should_stop:
+            return
+
+        if not self.is_ddp_backend:
+            message = "Trial was pruned at epoch {}.".format(epoch)
+            raise optuna.TrialPruned(message)
+        else:
+            # stop every ddp process if global rank 0 process decides to stop
+            trainer.should_stop = True
+            if trainer.is_global_zero:
+                self._trial.set_system_attr("pruned", True)
+                self._trial.set_system_attr("epoch", epoch)
```

Comment:
How about defining the key names at the top of the code? NSGA-II sampler employs such a convention, and I think it will prevent developers from adding minor bugs. In addition, the key name can be more descriptive such as  `ddp_pl:should_pruned` and `ddp_pl:epoch`

https://github.com/optuna/optuna/blob/ce6f856439ca2871f8a3d6d5bd91bcf6e6197fe1/optuna/samplers/_nsga2.py#L29-L33